Allow setting target subnet and static IP on failover

martenbohlin opened the pull request:
This pull request adds some configurability of the network interface when a failover is done.

The previous implementation finished once the initial setup was done and was not waiting for the replication to finish. But to configure the NIC we need to wait for the replication to finish before we can do that, in this pull request it will always wait for the replication even if no changes has to be done and I am not sure if that is desirable or not. Thoughts?

martenbohlin commented:
Hi, is there something I can do to increase the chance of this pull request to be accepted? If it is the dependency update that is the concern I can take a look and see if it is possible to implement this without the upgrade.

Thanks for maintaining a great product! /Mårten

martenbohlin commented:
I removed the upgrade of the recovery API since it did not merge cleanly with master. So I think you can remove the "dependencies" label now.

martenbohlin commented:
I do not think this PR was included in 2.9.0

Is it possible to get it into next release?

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -350,3 +534,62 @@ func resourceArmSiteRecoveryReplicatedVMDiskHash(v interface{}) int {
+		protectionContainerName := id.Path["replicationProtectionContainers"]
+		name := id.Path["replicationProtectedItems"]
+
+		client := meta.(*clients.Client).RecoveryServices.ReplicationMigrationItemsClient(resGroup, vaultName)
```
could we keep clients defied at the top of functions for consistency

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -350,3 +534,62 @@ func resourceArmSiteRecoveryReplicatedVMDiskHash(v interface{}) int {
+		if err != nil {
+			return nil, "", err
+		}
+		resGroup := id.ResourceGroup
```
```suggestion

		resGroup := id.ResourceGroup
```

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -350,3 +534,62 @@ func resourceArmSiteRecoveryReplicatedVMDiskHash(v interface{}) int {
+		}
+
+		// Find first disk that is not fully replicated yet
+		if a2aDetails, isA2a := resp.Properties.ProviderSpecificDetails.AsA2AReplicationDetails(); isA2a {
```
we should check that Properties/ProviderSpecificDetails are not nil

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -350,3 +534,62 @@ func resourceArmSiteRecoveryReplicatedVMDiskHash(v interface{}) int {
+			if a2aDetails.MonitoringPercentageCompletion != nil {
+				log.Printf("Waiting for Site Recover to replicate VM, %d%% complete.", *a2aDetails.MonitoringPercentageCompletion)
+			}
+			return resp, *a2aDetails.VMProtectionState, nil
```
we should check VMProtectionState is not nil too

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -350,3 +534,62 @@ func resourceArmSiteRecoveryReplicatedVMDiskHash(v interface{}) int {
+			}
+			return resp, *a2aDetails.VMProtectionState, nil
+		}
+		return resp, *resp.Properties.ReplicationHealth, nil
```
and check ReplicationHealth for nil

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -301,6 +473,18 @@ func resourceArmSiteRecoveryReplicatedItemRead(d *schema.ResourceData, meta inte
+			nicsOutput := make([]interface{}, 0)
+			for _, nic := range *a2aDetails.VMNics {
+				nicOutput := make(map[string]interface{})
+				nicOutput["source_network_interface_id"] = *nic.SourceNicArmID
+				nicOutput["target_static_ip"] = *nic.ReplicaNicStaticIPAddress
+				nicOutput["target_subnet_name"] = *nic.RecoveryVMSubnetName
```
we should be nil checking these dereferences

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -246,8 +315,110 @@ func resourceArmSiteRecoveryReplicatedItemCreate(d *schema.ResourceData, meta in
+			return fmt.Errorf("target_network_id must be set when a network_interface is configured")
+		}
+	}
+	var parameters = siterecovery.UpdateReplicationProtectedItemInput{
```
```suggestion

	var parameters = siterecovery.UpdateReplicationProtectedItemInput{
```

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -246,8 +315,110 @@ func resourceArmSiteRecoveryReplicatedItemCreate(d *schema.ResourceData, meta in
+			},
+		},
+	}
+	future, err := client.Update(ctx, fabricName, sourceProtectionContainerName, name, parameters)
```
```suggestion

	future, err := client.Update(ctx, fabricName, sourceProtectionContainerName, name, parameters)
```

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -161,14 +169,76 @@ func resourceArmSiteRecoveryReplicatedVM() *schema.Resource {
+	resGroup := d.Get("resource_group_name").(string)
+	vaultName := d.Get("recovery_vault_name").(string)
+
+	client := meta.(*clients.Client).RecoveryServices.ReplicationMigrationItemsClient(resGroup, vaultName)
```
could we keep client defs at the top of functions

katbyte commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -161,14 +169,76 @@ func resourceArmSiteRecoveryReplicatedVM() *schema.Resource {
+		return err1
+	}
+
+	err2 := updateReplication(d, client, meta)
```
This also means the resource will not return until replication is healthy where as before it would return before it does? what is the effect on the time to create the replication here? why do we need to update to finish the creation? could we document that in a comment

martenbohlin replied:
To configure the NIC we need to wait for the replication. Replication usually take a long time, so we could save time and only wait if a NIC configuration is present. But then it would behave different if you choose to configure the NIC or not. For my usecase it is good that the VM is replicated once terraform returns but I understand that is not always the case. So I am open for suggestions here.

martenbohlin commented on azurerm/internal/services/recoveryservices/resource_arm_site_recovery_replicated_vm.go:
```diff
@@ -161,14 +169,76 @@ func resourceArmSiteRecoveryReplicatedVM() *schema.Resource {
+	}
+}
+
+func netwrokInterfaceResource() *schema.Resource {
```
It is used both on line 177 in resourceArmSiteRecoveryReplicatedVM and line 
487 in resourceArmSiteRecoveryReplicatedItemRead

martenbohlin commented on azurerm/internal/services/recoveryservices/tests/resource_arm_site_recovery_replicated_vm_test.go:
```diff
@@ -207,12 +230,18 @@ resource "azurerm_site_recovery_replicated_vm" "test" {
   ]
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.Locations.Secondary, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.R ...
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.Locations.Secondary, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.R ...
```
Much better! Did know that syntax.

martenbohlin commented on azurerm/internal/services/recoveryservices/tests/resource_arm_site_recovery_replicated_vm_test.go:
```diff
@@ -246,6 +275,29 @@ func testCheckAzureRMSiteRecoveryReplicatedVmExists(resourceName string) resourc
 	}
 }
 
+func testCheckCorrectTargetSubnet(resourceName string, expectedTargetSubnet string) resource.TestCheckFunc {
```
Yes, you are correct. Removed that test.

martenbohlin commented:
@jackofallops I have fixed the typo you pointed out. I think I have fixed all of  @katbyte 's comments, but let me know if you have spotted something I have missed. Thanks

jackofallops commented:
Tests passing, failures are transient or intermittent unrelated items:
![image](https://user-images.githubusercontent.com/11830746/89633723-2d04ab80-d89c-11ea-99b1-f211068a3c20.png)
